compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java

```diff
@@ -960,12 +987,21 @@ public void visit(BLangInvocation.BLangActionInvocation actionInvocationExpr) {
     @Override
     public void visit(BLangTernaryExpr ternaryExpr) {
         ternaryExpr.thenExpr.accept(this);
-        boolean thenTaintedCheckResult = getObservedTaintedStatus();
+        Boolean thenTaintedCheckResult = this.taintedStatus;
         nonOverridingAnalysis = true;
         ternaryExpr.elseExpr.accept(this);
         nonOverridingAnalysis = false;
-        boolean elseTaintedCheckResult = getObservedTaintedStatus();
-        this.taintedStatus = thenTaintedCheckResult || elseTaintedCheckResult;
+        Boolean elseTaintedCheckResult = this.taintedStatus;
+        if (thenTaintedCheckResult == null && elseTaintedCheckResult == null) {
+            this.taintedStatus = null;
+        } else {
+            if (thenTaintedCheckResult == null) {
+                thenTaintedCheckResult = false;
+            } else if (elseTaintedCheckResult == null) {
+                elseTaintedCheckResult = false;
+            }
+            this.taintedStatus = thenTaintedCheckResult || elseTaintedCheckResult;
```

Comment:
This logic seems duplicated in several places. Shall we move them to a single method?

Reply:
This was changed with the enum change done and move to a method.